### PR TITLE
Add Realm models

### LINKED ISSUE #101: Generate realm models from model interfaces
Take into account #33.

### LINKED ISSUE #103: Convert models interfaces to concrete implementation for realm to consume

### PR DESCRIPTION
**Pull request by willsb**
Closes #101 
Closes #103 

Squash whenever

### CODE REVIEW

File: Toggl.PrimeRadiant.Realm/Models/RealmConstructors.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        public RealmClient(IDatabaseClient entity)
+            : this(entity as IClient)
+        {
+            IsDirty = entity.IsDirty;
+        }
+
+        public RealmClient(IClient entity)
+        {
+            Id = entity.Id;
+            WorkspaceId = entity. WorkspaceId;
```
Comment (amulware) [inline Toggl.PrimeRadiant.Realm/Models/RealmConstructors.cs]: Funny spaces.

File: Toggl.PrimeRadiant.Realm/Models/RealmConstructors.tt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        WriteLine( "        public bool IsDirty { get; set; }\n");
+        WriteLine($"        public {realmClassName}() {{ }}\n");
+        WriteLine($"        public {realmClassName}(IDatabase{className} entity)");
+        WriteLine($"            : this(entity as {interfaceName})");
+        WriteLine( "        {");
+        WriteLine( "            IsDirty = entity.IsDirty;");
+        WriteLine( "        }\n");
+        WriteLine($"        public {realmClassName}({interfaceName} entity)");
+        WriteLine( "        {");
+        WriteLine( "            Id = entity.Id;");
```
Comment (amulware) [inline Toggl.PrimeRadiant.Realm/Models/RealmConstructors.tt]: Why not use a multi-line string?

**Comment by amulware** [comment]
Couple of points:

1. I don't understand why partial classes are used. Are we not going to generate the models entirely?
 Especially since this claims to close #101 which specifies that we will generate the models from the interfaces.
2. Why are all the properties public-set? Is this needed for realm?

**Comment by willsb** [comment]
1 - We agreed on not generating everything because we might want to deal with model migrations manually. What I could do is have separate `.tt` files for each model and one for the constructors. Then, if a model requires migration, we deal with it specifically.  Agreed?

2 - Yup, Realm needs them. We'll deal with interfaces most of the times anyways, so it's fine.

**Comment by amulware** [comment]
One .tt for each model? Not sure that's really all that helpful then. I suppose in the end, since they implement the interface, we can be sure that the models are correct as long as they compile. Unless you have some other advantage of using generation here, I say we can leave it like this.

Migration will be a whole different problem that I don't wanna think about for the moment.

File: Toggl.PrimeRadiant.Realm/Models/RealmConstructors.tt (hunk 2 above)
Comment (willsb) [inline Toggl.PrimeRadiant.Realm/Models/RealmConstructors.tt]: Tried it, but the tooling for t4 is old and doesn't seem to support both multi-line and string interpolation. `$@""` breaks everything badly.

**Comment by amulware** [comment]
Somehow no tests were run for the last commit. Let's keep an eye on what's going on there.

**Comment by amulware** [comment]
According to bitrise, prime radiant does not build because of improper implementation of the model interfaces.

**Comment by amulware** [comment]
:+1:

**Pull request merged**